src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculator.java

```diff
@@ -0,0 +1,620 @@
+            return result.copy();
+        } else {
+            lastOverheadCounts.increase(index - lastOverheadCounts.index());
+            return lastOverheadCounts.copy();
+        }
+    }
+
+    /**
+     * Calculate the likelihoods given the list of alleles and the likelihood map.
+     *
+     * <p>This operation is <b>thread-unsafe</b>.</p>
+     *
+     * @param likelihoods the likelihood matrix all alleles vs all reads.
+     *
+     * @throws IllegalArgumentException if {@code alleleList} is {@code null} or {@code likelihoods} is {@code null}
+     *     or the alleleList size does not match the allele-count of this calculator, or there are missing allele vs
+     *     read combinations in {@code likelihoods}.
+     *
+     * @return never {@code null}.
+     */
+    public <A extends Allele> GenotypeLikelihoods genotypeLikelihoods(final LikelihoodMatrix<A> likelihoods) {
+        if (likelihoods == null) {
+            throw new IllegalArgumentException("the likelihood map cannot be null");
+        }
+
+        if (likelihoods.numberOfAlleles() != alleleCount) {
+            throw new IllegalArgumentException("mismatch between allele list and alleleCount");
+        }
+
+        final int readCount = likelihoods.numberOfReads();
+
+        ensureReadCapacity(readCount);
+
+        /// [x][y][z] = z * LnLk(Read_x | Allele_y)
+        final double[] readLikelihoodComponentsByAlleleCount
+                = readLikelihoodComponentsByAlleleCount(likelihoods);
+        final double[][] genotypeLikelihoodByRead = genotypeLikelihoodByRead(readLikelihoodComponentsByAlleleCount,readCount);
+        final double[] readLikelihoodsByGenotypeIndex = genotypeLikelihoods(genotypeLikelihoodByRead, readCount);
+        return GenotypeLikelihoods.fromLog10Likelihoods(readLikelihoodsByGenotypeIndex);
+    }
+
+    /**
+     * Calculates the final genotype likelihood array out of the likelihoods for each genotype per read.
+     *
+     * @param readLikelihoodsByGenotypeIndex <i>[g][r]</i> likelihoods for each genotype <i>g</i> and <i>r</i>.
+     * @param readCount number of reads in the input likelihood arrays in {@code genotypeLikelihoodByRead}.
+     * @return never {@code null}, one position per genotype where the <i>i</i> entry is the likelihood of the ith
+     *   genotype (0-based).
+     */
+    private double[] genotypeLikelihoods(final double[][] readLikelihoodsByGenotypeIndex, final int readCount) {
+        final double[] result = new double[genotypeCount];
+        final double denominator = readCount * log10[ploidy]; // instead of dividing each read likelihood by ploidy
+         // ( so subtract log10(ploidy) )  we multiply them all and the divide by ploidy^readCount (so substract readCount * log10(ploidy) )
+        for (int g = 0; g < genotypeCount; g++) {
+            final double[] likelihoodsByRead = readLikelihoodsByGenotypeIndex[g];
+            double s = - denominator;
+            for (int r = 0; r < readCount; r++) {
```

Comment:
This should be written using an array sum method in `MathUtils`.  If no method exists for summing the first `readCount` values of `likelihoodsByRead` it should be written.  This code would then become `result[g] = MathUtils.sum(likelihoodsByRead[g], readCount) - denominator`, which is much clearer.